src/parcsr_ls/par_cheby_device.c

```diff
@@ -0,0 +1,176 @@
+/******************************************************************************
+ * Copyright 1998-2019 Lawrence Livermore National Security, LLC and other
+ * HYPRE Project Developers. See the top-level COPYRIGHT file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ ******************************************************************************/
+
+/******************************************************************************
+ *
+ * Chebyshev setup and solve Device
+ *
+ *****************************************************************************/
+
+#include "_hypre_parcsr_ls.h"
+#include "_hypre_parcsr_mv.h"
+#include "float.h"
+
+#if defined(HYPRE_USING_CUDA)
+#include "_hypre_utilities.hpp"
+
+/**
+ * @brief Solve using a chebyshev polynomial on the device
+ *
+ * @param[in] A Matrix to relax with
+ * @param[in] f right-hand side
+ * @param[in] ds_data Diagonal information
+ * @param[in] coefs Polynomial coefficients
+ * @param[in] order Order of the polynomial
+ * @param[in] scale Whether or not to scale by diagonal
+ * @param[in] scale Whether or not to use a variant
+ * @param[in,out] u Initial/updated approximation
+ * @param[out] v Temp vector
+ * @param[out] v Temp Vector
+ */
+HYPRE_Int
+hypre_ParCSRRelax_Cheby_SolveDevice(hypre_ParCSRMatrix *A, /* matrix to relax with */
+                                    hypre_ParVector    *f, /* right-hand side */
+                                    HYPRE_Real         *ds_data,
+                                    HYPRE_Real         *coefs,
+                                    HYPRE_Int           order, /* polynomial order */
+                                    HYPRE_Int           scale, /* scale by diagonal?*/
+                                    HYPRE_Int           variant,
+                                    hypre_ParVector    *u, /* initial/updated approximation */
+                                    hypre_ParVector    *v /* temporary vector */,
+                                    hypre_ParVector    *r /*another temp vector */)
+{
+   hypre_CSRMatrix *A_diag = hypre_ParCSRMatrixDiag(A);
+   HYPRE_Real      *u_data = hypre_VectorData(hypre_ParVectorLocalVector(u));
+   HYPRE_Real      *f_data = hypre_VectorData(hypre_ParVectorLocalVector(f));
+   HYPRE_Real      *v_data = hypre_VectorData(hypre_ParVectorLocalVector(v));
+
+   HYPRE_Real *r_data = hypre_VectorData(hypre_ParVectorLocalVector(r));
+
+   HYPRE_Int i;
+   HYPRE_Int num_rows = hypre_CSRMatrixNumRows(A_diag);
+
+   HYPRE_Real  mult;
+   HYPRE_Real *orig_u;
+
+   HYPRE_Int cheby_order;
+
+   HYPRE_Real *tmp_data;
+
+   hypre_ParVector *tmp_vec;
+
+   /* u = u + p(A)r */
+
+   if (order > 4) order = 4;
+   if (order < 1) order = 1;
+
+   /* we are using the order of p(A) */
+   cheby_order = order - 1;
+
+   orig_u = hypre_CTAlloc(HYPRE_Real, num_rows, hypre_CSRMatrixMemoryLocation(A_diag));
+
```

Comment:
cudaMalloc is expensive. We should try to avoid doing it at every iteration. If temporary space is needed, consider allocating it outside.

Reply:
Handled now in `par_amg_setup.c` by using Ptemp/Rtemp when needed (ffaf12cba10e92fadc608004d0a59c58eac16453)